mcmcfit.py

```diff
@@ -1,871 +1,393 @@
-    # The first three columns are the data (x, y and y error)
-    # The next column is the CV flux
-    # The next columns are the flux from wd, bright spot, disc and donor
-    # Setup header for output file format
-    # outfile_header =
-
-    # Create a model from the first eclipses (eclipse 0) parameters
-    parNames = ['wdFlux_0', 'dFlux_0', 'sFlux_0', 'rsFlux_0', 'q', 'dphi',\
-        'rdisc_0', 'ulimb_0', 'rwd', 'scale_0', 'az_0', 'fis_0', 'dexp_0', 'phi0_0']
-    if complex:
-        parNames.extend(['exp1_0', 'exp2_0', 'tilt_0', 'yaw_0'])
-    # List of values obtained from input file using fromString function from mcmc_utils.py
-    parList = [Param.fromString(name, input_dict[name]) for name in parNames]
-    #parList = [Param.fromString(name, input_dict[name]) for name in parNames if Param.isValid]
-    # If fitting using GPs use GPLCModel, else use LCModel
-    if useGP:
-        model = GPLCModel(parList,complex,ampin_gp,ampout_gp,tau_gp)
-    else:
-        model = LCModel(parList,complex)
-
-    # pickle is used for parallelisation
-    # pickle cannot pickle methods of classes, so we wrap
-    # the ln_prior, ln_like and ln_prob functions here to
-    # make something that can be pickled
-    def ln_prior(parList):
-        model.pars = parList
-        return model.ln_prior()
-    def ln_like(parList,phi,y,e,width=None):
-        model.pars = parList
-        #return model.ln_like(phi,y,e,width)
-        lnlike = model.ln_like(phi,y,e,width)
-        if np.isfinite(lnlike):
-            return lnlike
+    comp_scat = bool(int(input_dict['comp_scat']))
+
+    # neclipses no longer strictly necessary, but can be used to limit the
+    # maximum number of fitted eclipses
+    try:
+        neclipses = int(input_dict['neclipses'])
+    except KeyError:
+        neclipses = -1
+
+    if use_gp:
+        # TODO: Impliment the GP version of the Eclipses.
+        # Read in GP params using fromString function from mcmc_utils.py
+        ampin_gp = Param.fromString('ampin_gp', input_dict['ampin_gp'])
+        ampout_gp = Param.fromString('ampout_gp', input_dict['ampout_gp'])
+        tau_gp = Param.fromString('tau_gp', input_dict['tau_gp'])
+
+    # Start by creating the overall Model. Gather the parameters:
+    core_par_names = ['rwd', 'dphi', 'q']
+    core_pars = [Param.fromString(name, s) for name, s in input_dict.items()
+                 if name in core_par_names]
+    # and make the model object with no children
+    model = LCModel('core', core_pars)
+
+    # Collect the bands and their params. Add them total model.
+    band_par_names = ['wdFlux', 'rsFlux']
+
+    # Get a sub-dict of only band parameters
+    band_dict = {}
+    for key, string in input_dict.items():
+        if np.any([key.startswith(par) for par in band_par_names]):
+            band_dict[key] = string
+
+    # Get a set of the bands we have in the input file
+    defined_bands = [key.split('_')[-1] for key in band_dict]
```

Comment:
will this still work if we decide not to lump all the g-band eclipses together, e.g if one band is called `g_0` and another is  `g_1`?

Reply:
This is a good point. Underscores would be forbidden inside labels, but the 'better' ways I can think of are kinda janky. Something like;
```
band_labels = [key.replace('band_', '') for key in band_dict]
```

But I think it would allow stuff like `g_hipercam_observations_taken_before_2017` or whatever, so long as it doesn't contain `band_`. An alternative is;
```
n_strip = len('band_')
band_labels = [key[:n_strip] for key in band_dict]
```

Thoughts?

Reply:
Why not use a different separator between ```class_type``` and ```label``` for a ```Node```? If you use an unusual separator here, like ```:``` or ```|``` then underscores would be valid in labels?

Reply:
That's probably more elegant. I ended up doing this:
`par, label = key.split('_')[0], '_'.join(key.split('_')[1:])`

But I think I'll swap to a version that has the option of being either a `:` or a pipe.